Sources/Atoms/Core/StoreContext.swift

```diff
@@ -336,43 +336,103 @@ private extension StoreContext {
         atom: Node,
         for key: AtomKey,
         newValue: Node.Loader.Value,
-        cache: AtomCache<Node>
+        oldValue: Node.Loader.Value
     ) {
-        let oldValue = cache.value
+        store.state.caches[key] = AtomCache(atom: atom, value: newValue)
 
-        store.state.caches[key] = mutating(cache) {
-            $0.value = newValue
+        // Check whether if the dependent atoms should be updated transitively.
+        guard atom._loader.shouldUpdateTransitively(newValue: newValue, oldValue: oldValue) else {
+            return
         }
 
-        guard atom._loader.shouldUpdate(newValue: newValue, oldValue: oldValue) else {
-            return
+        // Perform side effects first.
+        let state = getState(of: atom, for: key)
+        let context = AtomCurrentContext(store: self, coordinator: state.coordinator)
+        atom.updated(newValue: newValue, oldValue: oldValue, context: context)
```

Comment:
Q: Does this mean we're changing the timing of the side effects, and in case usage accidentally relies on the current order (performing side effects at the last of this method), it might get affected?

Reply:
Yes, before the change, it runs a side effect after finishing updating downstream, but now it runs immediately after updating its value.
In the first place, the order in which the atoms were updated was not guaranteed, so it's not possible that they were used in a way that depended on the update order. (Even if there were, it wouldn't work properly)

Reply:
> In the first place, the order in which the atoms were updated was not guaranteed, so it's not possible that they were used in a way that depended on the update order. (Even if there were, it wouldn't work properly)

Understood, thank you!

Reply:
@shingt 
Did you meant to approve or still reviewing?